An attempt was made to understand getting data directly from browser first

# Data Source
Website: Layoffs.fyi
Data Source: 
Airtable Shared View

Application ID:
app1PaujS9zxVGUZ4

View ID:
viwN3RMGptp84mfag

Share ID:
shroKsHx3SdYY0zeh


In [4]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

In [5]:
url="https://airtable.com/v0.3/view/viwN3RMGptp84mfag/readSharedViewData"
response = requests.get(url)
print("Status Code:", response.status_code)

Status Code: 403


In [6]:
import requests

url = "https://airtable.com/embed/app1PaujS9zxVGUZ4/shroKsHx3SdYYOzeh?backgroundColor=green&viewControls=on"

response = requests.get(url)

print(response.status_code)

200


In [7]:
print(response.text[:1000])

<!DOCTYPE html><html lang="en">
<head>
<meta name="referrer" content="no-referrer"/><meta name="robots" content="noindex"/><meta charSet="utf-8"/><meta name="theme-color" content="#ffffff"/><meta name="description" content="Explore the &quot;Layoffs.fyi  - Tech Layoffs Tracker&quot; view on Airtable."/><meta name="twitter:card" content="summary_large_image"/><meta name="twitter:site" content="@airtable"/><meta property="fb:app_id" content="1457397314575627"/><meta property="og:site_name" content="Airtable"/><meta property="og:title" content="Layoffs.fyi  - Tech Layoffs Tracker - Airtable"/><meta property="og:type" content="website"/><meta property="og:description" content="Explore the &quot;Layoffs.fyi  - Tech Layoffs Tracker&quot; view on Airtable."/><meta property="og:image" content="https://static.airtable.com/images/oembed/airtable_brand.png"/><meta http-equiv="x-ua-compatible" content="ie=edge"/><meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.

In [8]:
print(len(response.text))

105607


In [9]:
"Layoffs.fyi" in response.text

True

In [10]:
"app1PaujS9zxVGUZ4" in response.text

True

In [11]:
"shroKsHx3SdYYOzeh" in response.text

True

In [12]:
response.text.find("readSharedViewData")

4333

In [13]:
start = 4000
end = 5000

print(response.text[start:end])

ion(resolve, reject) {
setTimeout(
function() { reject('timeout'); },
330000);
});
var abortController;
var signal;
if (window.AbortController) {
abortController = new AbortController();
signal = abortController.signal;
}
window.__stashedPrefetch = {
metadata: {"method":"get","url":"\u002Fv0.3\u002Fview\u002FviwN3RMGptp84mfag\u002FreadSharedViewData","params":{"shouldUseNestedResponseFormat":true}},
urlWithParams: "\u002Fv0.3\u002Fview\u002FviwN3RMGptp84mfag\u002FreadSharedViewData?stringifiedObjectParams=%7B%22shouldUseNestedResponseFormat%22%3Atrue%7D&requestId=reqRAnymD0zRMApPb&accessPolicy=%7B%22allowedActions%22%3A%5B%7B%22modelClassName%22%3A%22view%22%2C%22modelIdSelector%22%3A%22viwN3RMGptp84mfag%22%2C%22action%22%3A%22readSharedViewData%22%7D%2C%7B%22modelClassName%22%3A%22view%22%2C%22modelIdSelector%22%3A%22viwN3RMGptp84mfag%22%2C%22action%22%3A%22getMetadataForPrinting%22%7D%2C%7B%22modelClassName%22%3A%22view%22%2C%22modelIdSelector%22%3A%22viwN3RMGptp84mfag%22%2C%22action

In [14]:
position = response.text.find("__stashedPrefetch")

print(position)

4229


In [15]:
print(response.text[position:position+3000])

__stashedPrefetch = {
metadata: {"method":"get","url":"\u002Fv0.3\u002Fview\u002FviwN3RMGptp84mfag\u002FreadSharedViewData","params":{"shouldUseNestedResponseFormat":true}},
urlWithParams: "\u002Fv0.3\u002Fview\u002FviwN3RMGptp84mfag\u002FreadSharedViewData?stringifiedObjectParams=%7B%22shouldUseNestedResponseFormat%22%3Atrue%7D&requestId=reqRAnymD0zRMApPb&accessPolicy=%7B%22allowedActions%22%3A%5B%7B%22modelClassName%22%3A%22view%22%2C%22modelIdSelector%22%3A%22viwN3RMGptp84mfag%22%2C%22action%22%3A%22readSharedViewData%22%7D%2C%7B%22modelClassName%22%3A%22view%22%2C%22modelIdSelector%22%3A%22viwN3RMGptp84mfag%22%2C%22action%22%3A%22getMetadataForPrinting%22%7D%2C%7B%22modelClassName%22%3A%22view%22%2C%22modelIdSelector%22%3A%22viwN3RMGptp84mfag%22%2C%22action%22%3A%22readSignedAttachmentUrls%22%7D%2C%7B%22modelClassName%22%3A%22row%22%2C%22modelIdSelector%22%3A%22rows%20*%5BdisplayedInView%3DviwN3RMGptp84mfag%5D%22%2C%22action%22%3A%22createDocumentPreviewSession%22%7D%5D%2C%22shareI

In [18]:
import re

match = re.search(r'urlWithParams:\s*"([^"]+)"', response.text)

print(match.group(1)[:500])

\u002Fv0.3\u002Fview\u002FviwN3RMGptp84mfag\u002FreadSharedViewData?stringifiedObjectParams=%7B%22shouldUseNestedResponseFormat%22%3Atrue%7D&requestId=reqRAnymD0zRMApPb&accessPolicy=%7B%22allowedActions%22%3A%5B%7B%22modelClassName%22%3A%22view%22%2C%22modelIdSelector%22%3A%22viwN3RMGptp84mfag%22%2C%22action%22%3A%22readSharedViewData%22%7D%2C%7B%22modelClassName%22%3A%22view%22%2C%22modelIdSelector%22%3A%22viwN3RMGptp84mfag%22%2C%22action%22%3A%22getMetadataForPrinting%22%7D%2C%7B%22modelClassN


In [19]:
import re

match = re.search(r'urlWithParams:\s*"([^"]+)"', response.text)

encoded_url = match.group(1)

decoded_url = encoded_url.encode().decode("unicode_escape")

full_url = "https://airtable.com" + decoded_url

print(full_url[:1000])

https://airtable.com/v0.3/view/viwN3RMGptp84mfag/readSharedViewData?stringifiedObjectParams=%7B%22shouldUseNestedResponseFormat%22%3Atrue%7D&requestId=reqRAnymD0zRMApPb&accessPolicy=%7B%22allowedActions%22%3A%5B%7B%22modelClassName%22%3A%22view%22%2C%22modelIdSelector%22%3A%22viwN3RMGptp84mfag%22%2C%22action%22%3A%22readSharedViewData%22%7D%2C%7B%22modelClassName%22%3A%22view%22%2C%22modelIdSelector%22%3A%22viwN3RMGptp84mfag%22%2C%22action%22%3A%22getMetadataForPrinting%22%7D%2C%7B%22modelClassName%22%3A%22view%22%2C%22modelIdSelector%22%3A%22viwN3RMGptp84mfag%22%2C%22action%22%3A%22readSignedAttachmentUrls%22%7D%2C%7B%22modelClassName%22%3A%22row%22%2C%22modelIdSelector%22%3A%22rows%20*%5BdisplayedInView%3DviwN3RMGptp84mfag%5D%22%2C%22action%22%3A%22createDocumentPreviewSession%22%7D%5D%2C%22shareId%22%3A%22shroKsHx3SdYYOzeh%22%2C%22applicationId%22%3A%22app1PaujS9zxVGUZ4%22%2C%22generationNumber%22%3A0%2C%22expires%22%3A%222026-06-18T00%3A00%3A00.000Z%22%2C%22signature%22%3A%228e5bed

In [20]:
test_response = requests.get(full_url)

print("Status:", test_response.status_code)
print("Content-Type:", test_response.headers.get("content-type"))
print("Length:", len(test_response.content))

Status: 401
Content-Type: application/json; charset=utf-8
Length: 3


In [21]:
import requests

session = requests.Session()

embed_url = "https://airtable.com/embed/app1PaujS9zxVGUZ4/shroKsHx3SdYYOzeh?backgroundColor=green&viewControls=on"

page = session.get(embed_url)

print("Embed Status:", page.status_code)

print("Cookies acquired:")
print(session.cookies.get_dict())

Embed Status: 200
Cookies acquired:
{'AWSALBTG': '8qe9nicHVtzoyBjc0EoCysZnfvwieMSfvOnJBJ4nNRFEhI69i2x9fmlmxuemAzZplwKljC8EqqDraOZCy5Zb17o/QHDAhwVp8ROHyvk6wW18OwF/KVxrU/0TNL8nbRRtlYV04MpExizbIt41cnsgoacEcyRU5o0FofH82SN/UyBlB9yv2h0=', 'AWSALBTGCORS': '8qe9nicHVtzoyBjc0EoCysZnfvwieMSfvOnJBJ4nNRFEhI69i2x9fmlmxuemAzZplwKljC8EqqDraOZCy5Zb17o/QHDAhwVp8ROHyvk6wW18OwF/KVxrU/0TNL8nbRRtlYV04MpExizbIt41cnsgoacEcyRU5o0FofH82SN/UyBlB9yv2h0=', '__Host-airtable-session': 'eyJzZXNzaW9uSWQiOiJzZXNvSWg3b0J6NW5kRXdPUCIsImNzcmZTZWNyZXQiOiJVYmVwNDc2dk13Vmx3eDdLeVp6M0pYQ1gifQ==', '__Host-airtable-session.sig': 'tsrSSF0_K65g5_6XpSM72djJ-JywSLTpJ--wX_4yduA', 'brw': 'brwOHIsqqUkP4KlCy', 'brwConsent': 'opt-in'}


In [22]:
api_response = session.get(full_url)

print("Status:", api_response.status_code)
print("Content-Type:", api_response.headers.get("content-type"))
print("Length:", len(api_response.content))

Status: 401
Content-Type: application/json; charset=utf-8
Length: 3


# Understanding the data from datasets



In [32]:
import pandas as pd

In [33]:
df1=pd.read_csv("../data/raw/layoffs_data.csv")

In [34]:
df1.head()

,Company,Location_HQ,Industry,Laid_Off_Count,Date,Source,Funds_Raised,Stage,Date_Added,Country,Percentage,List_of_Employees_Laid_Off
0,Oda,Oslo,Food,150.0,2024-06-05,https://techcrunch.com/2024/06/05/softbank-bac...,691.0,Unknown,2024-06-05 18:01:25,Norway,NaN,Unknown
1,Pagaya,Tel Aviv,Finance,100.0,2024-06-05,https://www.calcalistech.com/ctechnews/article...,2000.0,Post-IPO,2024-06-05 23:11:24,Israel,0.20,Unknown
2,Aleph Farms,Tel Aviv,Food,30.0,2024-06-05,https://www.calcalistech.com/ctechnews/article...,119.0,Unknown,2024-06-05 23:13:43,Israel,0.30,Unknown
3,MoonPay,Dover,Crypto,30.0,2024-06-05,https://www.theblock.co/post/298638/moonpay-la...,651.0,Unknown,2024-06-05 23:12:47,United States,0.10,Unknown
4,Yext,New York City,Marketing,NaN,2024-06-05,https://www.investing.com/news/economy-news/ye...,117.0,Post-IPO,2024-06-05 18:02:40,United States,0.12,Unknown


In [35]:
df1.shape

(3642, 12)

In [36]:
df1.info()

<class 'pandas.DataFrame'>
RangeIndex: 3642 entries, 0 to 3641
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Company                     3642 non-null   str    
 1   Location_HQ                 3642 non-null   str    
 2   Industry                    3642 non-null   str    
 3   Laid_Off_Count              2389 non-null   float64
 4   Date                        3642 non-null   str    
 5   Source                      3642 non-null   str    
 6   Funds_Raised                3252 non-null   float64
 7   Stage                       3642 non-null   str    
 8   Date_Added                  3642 non-null   str    
 9   Country                     3642 non-null   str    
 10  Percentage                  2342 non-null   float64
 11  List_of_Employees_Laid_Off  3642 non-null   str    
dtypes: float64(3), str(9)
memory usage: 341.6 KB


In [37]:
df1.columns

Index(['Company', 'Location_HQ', 'Industry', 'Laid_Off_Count', 'Date',
       'Source', 'Funds_Raised', 'Stage', 'Date_Added', 'Country',
       'Percentage', 'List_of_Employees_Laid_Off'],
      dtype='str')

In [38]:
df1.isnull().sum()

Company                          0
Location_HQ                      0
Industry                         0
Laid_Off_Count                1253
Date                             0
Source                           0
Funds_Raised                   390
Stage                            0
Date_Added                       0
Country                          0
Percentage                    1300
List_of_Employees_Laid_Off       0
dtype: int64

In [39]:
df1.describe()

,Laid_Off_Count,Funds_Raised,Percentage
count,2389.000000,3252.000000,2342.000000
mean,257.926329,798.569494,0.281216
std,829.432277,4758.393161,0.287294
min,3.000000,0.000000,0.000000
25%,40.000000,51.000000,0.100000
50%,80.000000,166.000000,0.170000
75%,181.000000,469.250000,0.300000
max,14000.000000,121900.000000,1.000000


In [40]:
df1["Date"].min(), df1["Date"].max()

('2020-03-11', '2024-06-05')

In [41]:
df1["Industry"].nunique()

31

In [42]:
df1["Country"].nunique()

63

In [43]:
df1["Stage"].nunique()

16

In [44]:
df1["Industry"].value_counts().head(10)

Industry
Finance           451
Retail            297
Healthcare        296
Transportation    226
Other             223
Food              213
Consumer          207
Marketing         178
Real Estate       151
Crypto            144
Name: count, dtype: int64

In [45]:
df1["Country"].value_counts().head(10)

Country
United States     2369
India              262
Canada             138
United Kingdom     126
Germany            103
Israel              87
Brazil              86
Australia           72
Singapore           49
Indonesia           38
Name: count, dtype: int64

In [46]:
df1["Stage"].value_counts()

Stage
Post-IPO          768
Unknown           618
Series B          429
Series C          405
Series D          316
Acquired          314
Series A          245
Series E          171
Seed              117
Series F          102
Private Equity     63
Series H           33
Series G           23
Subsidiary         21
Series J            9
Series I            8
Name: count, dtype: int64

In [47]:
df2 = pd.read_csv("../data/raw/layoffs.csv")

In [48]:
df2.head()

,company,location,total_laid_off,date,percentage_laid_off,industry,source,stage,funds_raised,country,date_added
0,SentinelOne,SF Bay Area,230.0,5/29/2026,0.08,Security,https://www.calcalistech.com/ctechnews/article...,Post-IPO,696.0,United States,5/29/2026
1,Rapyd,"Tel Aviv, Non-U.S.",NaN,5/28/2026,NaN,Finance,https://www.calcalistech.com/ctechnews/article...,Unknown,770.0,Israel,5/28/2026
2,NetApp,SF Bay Area,77.0,5/27/2026,NaN,Data,https://hoodline.com/2026/05/san-jose-mainstay...,Post-IPO,NaN,United States,5/28/2026
3,Lightricks,"Jerusalem, Non-U.S.",NaN,5/27/2026,NaN,Consumer,https://en.globes.co.il/en/article-lightricks-...,Series D,335.0,Israel,5/28/2026
4,Webflow,SF Bay Area,NaN,5/27/2026,NaN,Marketing,https://webflow.com/blog/evolving-webflow-for-...,Series C,335.0,United States,5/27/2026


In [49]:
df2.shape

(4428, 11)

In [50]:
df2.info()

<class 'pandas.DataFrame'>
RangeIndex: 4428 entries, 0 to 4427
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   company              4428 non-null   str    
 1   location             4427 non-null   str    
 2   total_laid_off       2900 non-null   float64
 3   date                 4428 non-null   str    
 4   percentage_laid_off  2787 non-null   float64
 5   industry             4426 non-null   str    
 6   source               4425 non-null   str    
 7   stage                4423 non-null   str    
 8   funds_raised         3916 non-null   float64
 9   country              4426 non-null   str    
 10  date_added           4428 non-null   str    
dtypes: float64(3), str(8)
memory usage: 380.7 KB


In [51]:
df2.columns

Index(['company', 'location', 'total_laid_off', 'date', 'percentage_laid_off',
       'industry', 'source', 'stage', 'funds_raised', 'country', 'date_added'],
      dtype='str')

In [52]:
df2["date"].min(), df2["date"].max()

('1/1/2023', '9/9/2024')

In [53]:
df2.isnull().sum()

company                   0
location                  1
total_laid_off         1528
date                      0
percentage_laid_off    1641
industry                  2
source                    3
stage                     5
funds_raised            512
country                   2
date_added                0
dtype: int64

In [54]:
df2["date"] = pd.to_datetime(df2["date"])

In [55]:
df2["date"].min(), df2["date"].max()

(Timestamp('2020-03-11 00:00:00'), Timestamp('2026-05-29 00:00:00'))

In [56]:
df2["industry"].nunique()

30

In [57]:
df2["country"].nunique()

66

In [58]:
df2["stage"].nunique()

16

In [59]:
df2["industry"].value_counts().head(10)

industry
Finance           527
Retail            353
Healthcare        327
Other             302
Consumer          301
Transportation    270
Food              250
Marketing         214
Crypto            165
Real Estate       165
Name: count, dtype: int64

# dataset layoffs 
4428 layoff events
66 countries
30 industries
16 company stages
2020-2026 coverage